File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.js
// bfs & visited
function solution(maps) {
  let result =[];
  let [H, W] = [maps.length, maps[0].length]; // height, width
  let visited = Array.from({ length: H }, () => Array(W).fill(0));
  let move = [[0,1], [0,-1], [1,0], [-1,0]]; // 우 좌 하 상

  // bfs함수
  function bfs(x,y) {
    let sum = 0; // 식량합
    let dq = [[x,y]]; // deque, 초기값 넣기
    sum += parseInt(maps[x][y]);
    visited[x][y] = 1; // 재방문 금지

    while (dq.length) {
      let [x, y] = dq.shift();
      for (let i=0; i<4; i++) { // 현 좌표에서 4방향 이동
        nowx = x + move[i][0], nowy = y + move[i][1]

        if (nowx>=0 && nowx<H && nowy>=0 && nowy<W) { // 1.범위 확인(맵 밖x)
          if (!visited[nowx][nowy] && maps[nowx][nowy] !== 'X') { // 2.갈수 있는가?
            visited[nowx][nowy] = 1; // 방문check
            sum += parseInt(maps[nowx][nowy]);
            dq.push([nowx,nowy]);
          }
        }
      };
    };
    // 총 식량 더하기
    result.push(sum);
  }

  for (let i=0; i<H; i++) {
    for (let j=0; j<W; j++) {
      if (!visited[i][j] && maps[i][j] !== 'X') { // bfs조건
        bfs(i,j)
      }
    }
  }

  return result.length === 0 ? [-1] : result.sort((a,b) => a-b)
}

console.log(solution(["X591X","X1X5X","X231X", "1XXX1"]))
console.log(solution(["XXX","XXX","XXX"]))

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
# 제한조건:
    # 3 ≤ maps의 길이 ≤ 100 / 3 ≤ maps[i]의 길이 ≤ 100
    # maps[i]는 'X' 또는 1 과 9 사이의 자연수로 이루어진 문자열
    # 지도는 직사각형 형태

# flood fill문제 같은데, 기초문제인듯
    # start는 0,0에서 시작한다 치고 언제 멈추는가? -> sx,sy가 맵끝(우측하단)에 도달할 때?
    # 이중포문, dfs가 아니라 bfs로 풀어야 함
from collections import deque

def solution(maps):
    width, height = len(maps[0]), len(maps)
    visited = [[0]*width for _ in range(height)]
    directions = [[0,1],[1,0],[0,-1],[-1,0]] # 우하좌상

    result = []
    for i in range(height):
        for j in range(width):
            sum = 0
            if maps[i][j] != 'X' and visited[i][j] == 0:
                q = deque()
                q.append([i,j])
                visited[i][j] = 10
                sum += int(maps[i][j])

                while q:
                    x, y = q.popleft()
                    for dx,dy in directions:
                        nowx, nowy = x+dx, y+dy
                        if 0 <= nowx < height and 0 <= nowy < width and maps[nowx][nowy] != 'X':
                            if not visited[nowx][nowy]:
                                visited[nowx][nowy] = 1
                                sum += int(maps[nowx][nowy])
                                q.append([nowx,nowy])
            
            if sum != 0:
                result.append(sum)
                        
                

    return sorted(result) if len(result) != 0 else [-1]


print(solution(["X591X","X1X5X","X231X", "1XXX1"])) # [1,1,27]
print(solution(["XXX","XXX","XXX"])) # [-1]